hypothesis-python/src/hypothesis/internal/conjecture/engine.py

```diff
@@ -832,6 +851,18 @@ def new_conjecture_data(self, prefix, max_length=BUFFER_SIZE, observer=None):
     def new_conjecture_data_for_buffer(self, buffer):
         return ConjectureData.for_buffer(buffer, observer=self.tree.new_observer())
 
+    @property
+    def max_shrinking_duration(self):
+        # If the shrinking phase takes too long, abort it early and print a warning.
+        # The specific deadline is somewhat adapative based on the deadline setting;
+        # defaults to 100 seconds but can go as low as 20 or as high as nine minutes.
+        #
+        # Many CI systems will kill a build after around ten minutes with no output,
+        # and of course appearing to hang isn't great for interactive use either -
+        # partially-shrunk examples are better than quitting with no examples!
+        deadline = (self.settings.deadline or timedelta(seconds=1)).total_seconds()
+        return max(20, min(MAX_SHRINKS * deadline, 9 * 60))
```

Comment:
I'm not sure the relationship to ```MAX_SHRINKS``` makes much sense here - a lot more than ```MAX_SHRINKS``` calls can be made during shrinking.

I kinda feel like this logic might be over-complicated. I'm happy to be argued with on this point but is there a good reason not to just have a fixed max shrinking time of, say, five minutes?

Reply:
The core idea was that 100 seconds seemed like a good default, but I'd want to scale it up if the `deadline` setting was higher than default to compensate for known-to-be-slow tests.

On the other hand "five minutes" is admirably easy to explain, and I don't have any concrete reason to prefer a more complicated calculation.  Will change this.